HLTrigger/Timer/interface/FastTimerService.h

```diff
@@ -86,690 +82,488 @@ class FastTimerService {
+  public:
+    PlotsPerProcess(ProcessCallGraph::ProcessType const&);
+    void reset();
+    void book(DQMStore::IBooker &, ProcessCallGraph const&, ProcessCallGraph::ProcessType const&,
+        double event_range, double event_resolution, double path_range, double path_resolution,
+        unsigned int lumisections, bool byls);
+    void fill(ProcessCallGraph::ProcessType const&, ResourcesPerJob const&, ResourcesPerProcess const&, unsigned int ls);
+
+  private:
+    // resources spent in all the modules of the (sub)process
+    PlotsPerElement           event_;
+    // resources spent in each path and endpath
+    std::vector<PlotsPerPath> paths_;
+    std::vector<PlotsPerPath> endpaths_;
+  };
 
+  class PlotsPerJob {
+  public:
+    PlotsPerJob(ProcessCallGraph const& job, std::vector<GroupOfModules> const& groups);
+    void reset();
+    void book(DQMStore::IBooker &, ProcessCallGraph const&, std::vector<GroupOfModules> const&,
+        double event_range, double event_resolution, double path_range, double path_resolution,
+        double module_range, double module_resolution, unsigned int lumisections,
+        bool bymodule, bool byls);
+    void fill(ProcessCallGraph const&, ResourcesPerJob const&, unsigned int ls);
+
+  private:
+    // resources spent in all the modules of the job
+    PlotsPerElement              event_;
+    // resources spent in the highlighted modules
+    std::vector<PlotsPerElement> highlight_;
+    // resources spent in each module
+    std::vector<PlotsPerElement> modules_;
+    // resources spent in each (sub)process
+    std::vector<PlotsPerProcess> processes_;
   };
 
-  // process descriptions
-  std::vector<ProcessDescription>               m_process;
 
-  // description of the luminosity axes
-  std::vector<LuminosityDescription>            m_dqm_luminosity;
+  // keep track of the dependencies among modules
+  ProcessCallGraph callgraph_;
 
-  // stream data
-  std::vector<StreamData>                       m_stream;
+  // per-stream information
+  std::vector<ResourcesPerJob>  streams_;
+  std::vector<PlotsPerJob>      stream_plots_;
 
   // summary data
-  std::vector<Timing>                           m_run_summary;                  // whole event time accounting per-run
-  Timing                                        m_job_summary;                  // whole event time accounting per-run
-  std::vector<std::vector<TimingPerProcess>>    m_run_summary_perprocess;       // per-process time accounting per-job
-  std::vector<TimingPerProcess>                 m_job_summary_perprocess;       // per-process time accounting per-job
-  std::mutex                                    m_summary_mutex;                // synchronise access to the summary objects across different threads
+  ResourcesPerJob               job_summary_;                   // whole event time accounting per-job
+  std::vector<ResourcesPerJob>  run_summary_;                   // whole event time accounting per-run
```

Comment:
No, this has as many elements as `edm::service::SystemBounds::maxNumberOfConcurrentRuns()`, and is indexed via `edm::GlobalContext::runIndex()` or `edm::StreamContext::runIndex()` .

My assumption is that there are never more than `maxNumberOfConcurrentRuns()` "in flight" at any given moment, and that any access happens between `preGlobalBeginRun` and `postGlobalEndRun`.